pyccel/parser/semantic.py

```diff
@@ -870,6 +904,92 @@ def _handle_function(self, expr, func, args, **settings):
+                    self._infere_type(a, **settings)
+            for a in kwargs.values():
+                if getattr(a,'dtype',None) == 'tuple':
+                    self._infere_type(a, **settings)
+
+            try:
+                new_expr = func(*args, **kwargs)
+            except TypeError:
+                errors.report(UNRECOGNISED_FUNCTION_CALL,
+                        symbol = expr,
+                        severity = 'fatal')
+
+            return new_expr
+        else:
+            if isinstance(func, FunctionDef) and len(args) > len(func.arguments):
+                errors.report("Too many arguments passed in function call",
+                        symbol = expr,
+                        severity='fatal')
+            # TODO : type check the NUMBER OF BLOCKS 'numBlocks' and threads per block 'tpblock'
+            if not isinstance(expr.numBlocks, LiteralInteger):
+                # expr.numBlocks could be invalid type, or PyccelSymbol
+                if isinstance(expr.numBlocks, PyccelSymbol):
+                    numBlocks = self.get_variable(expr.numBlocks)
+                    if not isinstance(numBlocks.dtype, NativeInteger):
+                        errors.report(INVALID_KERNEL_CALL_BP_GRID,
+                        symbol = expr,
+                        severity='error')
+                else:
+                    errors.report(INVALID_KERNEL_CALL_BP_GRID,
+                        symbol = expr,
+                        severity='error')
+            if not isinstance(expr.tpblock, LiteralInteger):
+                # expr.tpblock could be invalid type, or PyccelSymbol
+                if isinstance(expr.tpblock, PyccelSymbol):
+                    tpblock = self.get_variable(expr.tpblock)
+                    if not isinstance(tpblock.dtype, NativeInteger):
+                        errors.report(INVALID_KERNEL_CALL_TP_BLOCK,
+                        symbol = expr,
+                        severity='error')
+                else:
+                    errors.report(INVALID_KERNEL_CALL_TP_BLOCK,
+                        symbol = expr,
+                        severity='error')
+            new_expr = KernelCall(func, args, expr.numBlocks, expr.tpblock, self._current_function)
+
+            for a in new_expr.args:
+                if a is None:
+                    errors.report("Too few arguments passed in function call",
+                        symbol = expr,
+                        severity='error')
+                elif isinstance(a.value, Variable) and a.value.on_stack:
+                    errors.report("A variable allocated on the stack can't be passed to a Kernel function",
+                        symbol = expr,
+                        severity='error')
+            if isinstance(func, FunctionDef):
+                self._check_argument_compatibility(new_expr.args, func.arguments,
+                        expr, func.is_elemental)
+            return new_expr
```

Comment:
This seems to duplicate lots of code from `_handle_function` is there no way to collect all the differences together and call `_handle_function` directly?